test/sparse/sparse.jl

```diff
@@ -945,886 +1004,951 @@ function test_getindex_algs{Tv,Ti}(A::SparseMatrixCSC{Tv,Ti}, I::AbstractVector,
+
+        SA = [sprand(M, N, d) for d in [1., 0.1, 0.01, 0.001, 0.0001, 0.]]
         for S in SA
             res = Any[1,2,3]
-            times = Float64[0,0,0]
-            best = [typemax(Float64), 0]
             for searchtype in [0, 1, 2]
-                gc()
-                tres = @timed test_getindex_algs(S, I, J, searchtype)
-                res[searchtype+1] = tres[1]
-                times[searchtype+1] = tres[2]
-                if best[1] > tres[2]
-                    best[1] = tres[2]
-                    best[2] = searchtype
-                end
+                res[searchtype+1] = test_getindex_algs(S, I, J, searchtype)
             end
 
-            if debug
-                @printf(" %7d | %7d | %4.2e | %4.2e | %4.2e | %s\n", round(Int,nnz(S)/S.n), length(I), times[1], times[2], times[3],
-                            (0 == best[2]) ? "binary S" : (1 == best[2]) ? "binary I" : "linear")
-            end
-            if res[1] != res[2]
-                println("1 and 2")
-            elseif res[2] != res[3]
-                println("2, 3")
-            end
             @test res[1] == res[2] == res[3]
         end
     end
+
+    let M = 2^14, N=2^4
+        I = randperm(M)
+        J = randperm(N)
+        Jsorted = sort(J)
+
+        SA = [sprand(M, N, d) for d in [1., 0.1, 0.01, 0.001, 0.0001, 0.]]
+        IA = [I[1:round(Int,n)] for n in [M, M*0.1, M*0.01, M*0.001, M*0.0001, 0.]]
+        debug = false
+        if debug
+            @printf("         |         |         |        times        |        memory       |\n")
+            @printf("    S    |    I    |    J    |  sorted  | unsorted |  sorted  | unsorted |\n")
+        end
+        for I in IA
+            Isorted = sort(I)
+            for S in SA
+                gc()
+                ru = @timed S[I, J]
+                gc()
+                rs = @timed S[Isorted, Jsorted]
+                if debug
+                    @printf(" %7d | %7d | %7d | %4.2e | %4.2e | %4.2e | %4.2e |\n", round(Int,nnz(S)/S.n), length(I), length(J), rs[2], ru[2], rs[3], ru[3])
+                end
+            end
+        end
+    end
 end
 
-let M = 2^8, N=2^3
-    Irand = randperm(M)
-    Jrand = randperm(N)
-    I = sort([Irand; Irand; Irand])
-    J = [Jrand; Jrand]
+@testset "Random sparse bounds checking" begin
+    let S = sprand(10, 10, 0.1)
+        @test_throws BoundsError S[[0,1,2], [1,2]]
+        @test_throws BoundsError S[[1,2], [0,1,2]]
+        @test_throws BoundsError S[[0,2,1], [1,2]]
+        @test_throws BoundsError S[[2,1], [0,1,2]]
+    end
+end
 
-    SA = [sprand(M, N, d) for d in [1., 0.1, 0.01, 0.001, 0.0001, 0.]]
-    for S in SA
-        res = Any[1,2,3]
-        for searchtype in [0, 1, 2]
-            res[searchtype+1] = test_getindex_algs(S, I, J, searchtype)
-        end
+# Test that sparse / sparsevec constructors work for AbstractMatrix subtypes
+@testset "sparse / sparsevec AbstractMatrix constructors" begin
+    let D = Diagonal(ones(10,10)),
```

Comment:
If you mean beginning a testset with let instead of begin/for then I don't believe it is possible.

```
@testset "cats" let
end
ERROR: Expected begin/end block or for loop as argument to @testset
```